More fixes to the Screenshots UI tests and scripts

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: 392-gh-woocommerce/woomobile-private
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR adds some further improvements to the UI tests used for screenshots in addition to the Fastfile script, this is a continuation of the work started by Hafiz in previous PRs.
I'll post some details around the different changes in the comments below, so please make sure you read them.

## Steps to reproduce
1. Use the following command to build and generate screenshots:

```
bundle exec fastlane screenshots languages:ar-SA,en-US
```
2. Generate promo screenshots using these commands:
```
// If needed
brew install imagemagick automattic/build-tools/drawText

bundle install --with screenshots
bundle exec fastlane download_app_store_translations
bundle exec fastlane create_promo_screenshots
```

I'm limiting the generation to two languages, just to confirm that it works for both the source language and another translated one.

## Testing information
I tested the above.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hichamboushaba committed:
Scroll to top after returning from payments screen

Otherwise the order summary label could be hidden in the scrolled out area, and would lead to failure.

hichamboushaba committed:
Interact with the app before taking the first screenshot

We do this by using `TabNavComponent().goToMyStoreScreen()` instead of directly referencing `MyStoreScreen()`.

addUIInterruptionMonitor's callback is not called until we interact with the app, in the past, we used to use `app.tap()` for this, but since it was removed, the permission dialog would be still visible on the first screenshot. When I attempted re-introducing it back, I noticed that it caused issues on tablets by opening some unwanted screens, so I used `TabNavComponent().goToMyStoreScreen()`, it plays the same role now, as it also allows to interact with the app, and also solves the issue on tablets.

hichamboushaba committed:
Revert "Update the translation filename used in promo generating to the correct filenames."

This reverts commit 8cac92b4cec20c1609b9498ad42d612ff2085c3a.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Make sure the Arabic locale used for screenshots matches the expected one for promo generation

hichamboushaba committed:
Align screenshot text files naming between English and other locales

hichamboushaba commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -81,62 +81,20 @@ public final class CoreDataManager: StorageManagerType {
         return backgroundContext
     }()
 
+    private let queue = DispatchQueue(label: "persistentContainerQueue")
```
These changes will be dropped from the branch before merging the PR, they are needed now until the PR #14534 is merged.

hichamboushaba commented on WooCommerce/UITestsFoundation/Screens/Orders/PaymentMethodsScreen.swift:
```diff
@@ -35,6 +35,12 @@ public final class PaymentMethodsScreen: ScreenObject {
     @discardableResult
     public func goBackToOrderScreen() throws -> SingleOrderScreen {
         dismissButton.tap()
+
+        let orderDetailTableView = app.tables["order-details-table-view"]
+        // On smaller screen, the Summary area might not be visible anymore. We need to bring it back so when returning
+        // back from Payment Details, it can be detected.
+        orderDetailTableView.swipeDown()
+
```
This code was moved from `goBackToOrdersScreen` method in `SingleOrderScreen` to here, it was added in https://github.com/woocommerce/woocommerce-ios/pull/14335, but from my testing (and my understanding), the logic should be executed before that, because we'll need the order details's summary view to be visible before getting to `SingleOrderScreen`, without this change, the UI tests failed constantly on the iPhone 8 Plus.

hichamboushaba commented on WooCommerce/WooCommerceScreenshots/WooCommerceScreenshots.swift:
```diff
@@ -43,38 +43,38 @@ class WooCommerceScreenshots: XCTestCase {
         // My Store
-        .dismissTopBannerIfNeeded()
-        .thenTakeScreenshot(named: "order-dashboard")
+        try TabNavComponent()
+            .goToMyStoreScreen()
```
As explained in the commit a8d6044bacbadcd372bb55fc623b3a2692089b8c, we need to ensure that we interact with the app before taking any screenshots, otherwise in the first screenshot, the notifications permission dialog would be still visible, in the past, this interaction used `app.tap()`, but as Hafiz found out, it was causing some issues, and he removed it in this PR https://github.com/woocommerce/woocommerce-ios/pull/14492, so to solve both issues, I updated the logic to force tapping on the Dashboard bottom navigation icon.

hichamboushaba commented on fastlane/Fastfile:
```diff
@@ -25,7 +25,7 @@ SCREENSHOTS_DIR =  File.join(FASTLANE_DIR, 'screenshots')
 WORKSPACE_PATH = File.join(PROJECT_ROOT_FOLDER, 'WooCommerce.xcworkspace')
 BUILDKITE_RELEASE_PIPELINE = 'release-builds.yml'
-IOS_LOCALES = %w[ar de-DE en-US es-ES fr-FR he id it ja ko nl-NL pt-BR ru sv tr zh-Hans zh-Hant].freeze
+IOS_LOCALES = %w[ar-SA de-DE en-US es-ES fr-FR he id it ja ko nl-NL pt-BR ru sv tr zh-Hans zh-Hant].freeze
```
The promo screenshots generation expect `ar-SA` and not just `ar`, and since everything else (metadata handling and also fastlane [upload_to_app_store lane](https://docs.fastlane.tools/actions/deliver/)) expects `ar-SA`, I updated here for the screenshot generation step as well.

hichamboushaba commented on fastlane/Fastfile:
```diff
@@ -1083,7 +1083,7 @@ platform :ios do
     ].each do |filename|
       source = File.join(Dir.pwd, 'appstoreres', 'metadata', 'source', filename)
-      destination = File.join(en_us_path, filename)
+      destination = File.join(en_us_path, filename.gsub('promo_', 'app_store_'))
```
We rename the English screenshot title to `app_store_screenshot_%` like the other locales.

hichamboushaba commented on fastlane/screenshots.json:
```diff
@@ -63,39 +63,39 @@
 			"background": "#674399",
 			"screenshot": "iPhone 11 Pro Max-1-dark-order-dashboard.png",
-			"text": "metadata/{locale}/promo_screenshot_1.txt"
+			"text": "metadata/{locale}/app_store_screenshot_1.txt"
```
I reverted the commit 8cac92b4cec20c1609b9498ad42d612ff2085c3a that was added in the PR https://github.com/woocommerce/woocommerce-ios/pull/14492, there was just a misunderstanding I think, as the script and config all expects `app_store_screenshot_%`, the cause of the misunderstanding is that for English there was an issue, and now it should be fixed by the change mentioned in the comment ☝️

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>s...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan reviewed (approved):
Thanks for working on this, Hicham!

Your comments helped a lot to get context about the changes. This is excellent work navigating the UI tests and Fastlane scripts and making fixes. 🎖️ 

---

I could not verify the testing steps successfully on my machine. 

`bundle exec fastlane screenshots languages:ar-SA,en-US` failed with the following error,

```
[10:16:51]: Called from Fastfile at line 835
[10:16:51]: ```
[10:16:51]:     833:	    xcversion
[10:16:51]:     834:	
[10:16:51]:  => 835:	    scan(
[10:16:51]:     836:	      workspace: WORKSPACE_PATH,
[10:16:51]:     837:	      scheme: SCREENSHOTS_SCHEME,
[10:16:51]: ```
[10:16:51]: Error building/testing the application. See the log above.
```

The error looks like something to do with my Xcode path. We can work on this in a Slack thread. 

---

I am approving this PR to unblock development as fixing my local config isn't a blocker.

hichamboushaba commented:
Thanks @selanthiraiyan for the review.
> The error looks like something to do with my Xcode path. We can work on this in a Slack thread.

Yes, please share with me the full log and we can try to work on it together.

--------

For the PR itself, I rebased it on `trunk` and removed the commit referenced [here](https://github.com/woocommerce/woocommerce-ios/pull/14539#discussion_r1862422121), now we can merge this.
